docs/zkapps/tutorials/06-offchain-storage.mdx

```diff
@@ -0,0 +1,390 @@
+---
+title: "Tutorial 6: Offchain Storage"
+hide_title: true
+sidebar_label: "Tutorial 6: Offchain Storage"
+---
+
+:::info
+
+Please note that zkApp programmability is not yet available on Mina Mainnet, but
+zkApps can now be deployed to Berkeley Testnet.
+
+:::
+
+# Tutorial 6: Offchain Storage
+
+## Overview
+
+In our last tutorial, we ended by learning about merkle trees, and how we can use them to represent large amounts of data with a single hash.
+
+In this tutorial, we will share a library and pattern for working with storing merkle trees offchain, with only the root of the merkle tree stored on chain.
+
+:::caution
+
+This Library is ready for prototyping and building toy applications, but needs modification to be production ready. If you are interested in helping make that happen, see the [further development](#further-development) section below.
+
+:::
+
+### Why Offchain Storage?
+
+When building an application we'll just use locally and for testing, its fine to just build and store a merkle root locally. 
+
+However, when building a complete, distributed zkApp, we need more than this. All users interacting with our zkApp need to be able to get the latest state, and modify it.
+
+If this isn't the case, and someone can modify the tree without sharing their modification, the lack of data availability for the new version means that further writes, by other parties that don't have access to the new tree, become impossible.
+
+The solution here is to have a server, that anyone can run, that stores the data. If the server is being run as expected, it provides a guarantee that any updates to the merkle root reflected onchain will be stored on the server for anyone to retrieve.
+
+### Offchain Storage and Decentralization
+
+This solution works well as long as the server is being run as expected. This is reasonable when the users of the application and the developer both trust the server the zkApp is using.
+
+While this is useful for prototyping applications, and even putting many applications into production, its limited from a decentralization standpoint.
+
+In the short-run, we expect the following solutions for data availability:
+
+* Single-server offchain storage (the solution presented here).
+* Multi-server offchain storage - with guarantees as long as a majority of servers behave honestly (Grant available; see below)
+* Modular blockchain offchain storage (Several dependencies, but under exploration)
+
+In the long run, a feature is being considered for a future hard fork to add purchasable data storage with data availability guarantees to Mina itself. However, even then, putting more data on chain burdens the network, so it is likely that putting large amounts of data on chain will be relatively expensive. Given that it is expected that the short-run solutions presented above will still be relevant and useful.
```

Comment:
Suggestion for clarity:

> In the long run, Mina may even add the ability for zkApps to purchase or rent additional fields of on-chain storage, which would provide data availability (identical?) guarantees similar to existing on-chain storage. This would require a future hard fork. However, even then...

`Given that it is` > `Given this, it is`